### ISSUE TITLE
Revalidate perf of batch/server GC settings for fsc, fsi on .NET Framework and .NET Core

### ISSUE THREAD

**Issue by dsyme** [dotnet/fsharp#13739]
This is a tracking issue to make sure we revalidate the settings that we use for GC on both .NET Framework and .NET 6/7.

* gcServer
* LatencyMode.Batch

So the current situation is

* .NET SDK compiler: Server off, Batch on
* .NET Framework compiler: Server on, Batch on

**Comment by dsyme** [comment]
Here is the matrix on my machine for compiler-compiling-the-compiler, based on wallclock timings (they seem to vary about +/- 4%)

| .NET Version | GCBatch | GCServer | Parse | Typecheck | Optimize | Total | Notes |
|----:|----:|----:|----:|----:|----:|----:|:-----|
| .NET 472 32bit | no | no | 11.6 |  20.5 |  16.2 | 60.2 |  |
| .NET 472 32bit | yes | no | 11.6 | 20.8 | 16.4 | 60.4 |  |
| .NET 472 32bit | no | yes | 14.1 | 30.2 |  23.3 | 84.8 |  |
| .NET 472 32bit | yes | yes | 14.6 | 28.8  | 26.2 |  88.3 | This is what we ship today (though 64-bit tools are default) |
| | | | | | | | |
| .NET 472 64bit | no | no | 13.5 | 25.5 | 17.8  | 69.8 |  |
| .NET 472 64bit | yes | no | 14.0 | 25.4 | 18.6 | 71.8 |  |
| .NET 472 64bit | no | yes | 14.8 | 34.7 | 26.2  | 88.2 |  |
| .NET 472 64bit | yes | yes | 13.9 | 33.7 | 26.5 | 93.3 | This is what we ship  |
| | | | | | | | |
| .NET 6 | no | no | 6.9 | 22.3 | 15.0 | 54.7 |  |
| .NET 6 | yes | no | 7.8 | 22.8 | 15.1 | 56.1  | This is what we shipped in .NET 6 | 
| .NET 6 | no | yes | 7.0 | 31.1 | 20.9 | 70.6 |  | 
| .NET 6 | yes | yes | 6.7 | 31.9 | 20.9 | 72.9 | Server+batch causes significant slowdown |  
| | | | | | | | |
| .NET 7 p7 | no | no | 6.8 | 23.2 | 15.2 | 55.9 |   | 
| .NET 7 p7 | yes | no | -- | -- | -- | -- | This is what we ship today. Not run due to https://github.com/dotnet/runtime/issues/74286 | 
| .NET 7 p7 | no | yes | -- | -- | -- | -- | Not run due to a surprising NRE when GCServer enabled | 
| .NET 7 p7 | yes | yes | 7.4 | 33.0 | 23.5 | 77.2 | | 

Please don't compare between groups as unclear if ngen and crossgen are applied.

In short "no, no" is always the best.
* Server mode is a significant loss and we should never have turned it on for the command-line compiler for .NET Framework (and should stay off for .NET Core)
* Batch mode doesn't make much difference, but it seems it is no longer worth it, ever.  I assume this is because our compiler is now using more concurrency, or something like that.

**Comment by dsyme** [comment]
Hmmm  I think I'm wrong about ServerGC being slower. `--times` is showing as slower for overall CPU time reported but the server GC is actually faster on wall time. 

To see this, I took logs using MSBuild `-verbosity d` which includes a processor graph at the end with actual times for each process launched by the build and also shows core utilization. It's pretty wild though - according to `--times`  turning off Server GC takes total CPU time consumed by F# compiler in the build from ~400s to ~300s.   But the actual end-to-end time goes up by 10% or so.

Anyway looks like we need to change our timing methodology a little, certainly don't rely on `--times`.  End-to-end wallclock times now look more reliable.  

This also means we should probably turn ON server GC for the compiler in the .NET SDK.

**Comment by dsyme** [comment]
In more detail, I did some more validation on wether Server GC is a gain or a loss.

In https://github.com/dotnet/fsharp/issues/13739 I had validated manually that turning gcServer off improved compile time for compiling `FSharp.Compiler.Service.dll` alone. However tried to validate a perf improvement on end-to-end build of the repo (when the Bootstrap compiler uses .NET Framework).  Surprisingly I didn't see this

End-to-end build timings for `.\build -c Release`
* With gcServer: 139s
* Without gcServer: 146s

This surprised me as the overall compile time is quite heavily dominated by compiling `FSharp.Compiler.Service.dll`.   Digging deeper into the logs for the end to end build, `--times` indeed reports the same as I saw - that Server GC is slower for FSharp.Compiler.Service.dll:

    FSharp.Compiler.Service.dll:
        With Server GC:             TIME: 85.0 Delta:  9.1 Mem: 571 G0:  12 G1:  7 G2:  3 [Write .NET Binary]
        Without Server GC:         TIME: 61.0 Delta:  4.0 Mem: 784 G0: 121 G1: 107 G2:  0 [Write .NET Binary]

Adding `--times` to all compilations and analysing the entire build log reveals we indeed spend much much less CPU time in the F# compiler overall when Server GC is off:

    grep "Write .NET Binary" /mnt/c/GitHub/dsyme/with.log | col2 | paste -sd+ - | bc
    428.7
    grep "Write .NET Binary" /mnt/c/GitHub/dsyme/without.log | col2 | paste -sd+ - | bc
    320.2

This is a huge difference - so I was really confused why this doesn't translate to better end-to-end build times.   The memory usage reported by `--times` is also reasonable:

    grep "Write .NET Binary" /mnt/c/GitHub/dsyme/with.log | col6 | paste -sd+ - | bc
    10972
    grep "Write .NET Binary" /mnt/c/GitHub/dsyme/without.log | col6 | paste -sd+ - | bc
    12515

Then I looked at the actual end-to-end process times and CPU core utilizations reported by `.\build -verbosity d`.   This showed that the actual wall-clock execution times for processes were substantially lower with Server GC, to take one example:

```
With:            637966065343685385:   x       x       x       x       x       |       x       |       |       x       |       x       x       x       x       x       x       |       x       x        23.716s     106.233s ++++ (scale 1:100)

Without:          637966067913295762:   |       x       x       x       x       |       x       |       x       x       x       x       x       |       |       x       |       x       |       647      31.264s     115.905s ++++++ (scale 1:100)
```

Here the "with server GC" is using 15 cores and is 8% faster, while "without server GC" is using 12 cores.

So, in short, don't trust `--times`  which reports `Process.GetCurrentProcess().UserProcessorTime.TotalSeconds` and instead look at either wall-clock or these individual process reports.  I'll adjust `--times` to report wall-clock time.

**Comment by dsyme** [comment]
Here's our latest `main` compiler running on .NET 6, compiling FSharp.Compiler.Service.dll, batch mode turned off.
```
Without server GC,    Total CPU = 56.2     Wallclock = 50.4
With    server GC,    Total CPU = 78.6     Wallclock = 45.1
```
So wallclock gain is 10% with Server GC.

**Comment by dsyme** [comment]
Closing as done.  We need to enable Server mode by default, and have removed use of Batch
